### PR TITLE
[Feature] Edge Feature Support in GNN and base models

### PR DESCRIPTION
**Pull request by zhjwy9343**
*Issue #, if available:*

*Description of changes:*
This is the 4th PR for edge feature support. This PR include:
1. Modify `GSgnnNodeModel`, `GSgnnEdgeModel`, and `GSgnnLinkPredictionModel` forward and predict functions to use edge feature.
2. Modify `compute_embed_step()` and `do_mini_batch_inference` method in `GSgnnModel` to use edge features.
3. Add a `is_using_edge_feature` method in `GraphConvEncoder` for other classes to check if the encoder uses edge feature.
4. Add test cases in unit test.

**Backward compatibility breaking includes:**
1. Changes `input_feats` to `input_nfeats` in `compute_embed_step()` argument list;
2. Add a new argument `input_efeats_list` in `compute_embed_step()` argument list to support edge feature.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/unit-tests/test_embed.py [hunk 1]
```diff
@@ -191,6 +191,25 @@ def test_input_layer4(dev):
...
+    assert len(edge_input_layer.input_projs) == 0
+    edge_input_layer = edge_input_layer.to(dev)
+
+    block_edge_feat_list = [{}, {}]
+    embed = edge_input_layer(block_edge_feat_list)
+    assert len(embed) == len(block_edge_feat_list)
+    assert any(embed) == False
+
+    edge_feat_size = get_edge_feat_size(g, {('n0', 'r0', 'n1'): ['feat'], \
+                                            ('n0', 'r1', 'n1'): ['feat']})
```
Comment (classicsong) [inline tests/unit-tests/test_embed.py:206]: The edge_feat_size here is [{}, {}]?

File: tests/unit-tests/test_gnn.py [hunk 2]
```diff
@@ -748,24 +759,54 @@ def test_gat_node_prediction(device):
...
     gnn_encoder = RelationalGCNEncoder(g, 4, 4,
                                        num_bases=2,
                                        num_hidden_layers=1,
+                                       edge_feat_name=edge_feat_name,
                                        dropout=0,
                                        use_self_loop=True)
     model.set_gnn_encoder(gnn_encoder)
     model.set_decoder(MLPEdgeDecoder(model.gnn_encoder.out_dims,
-                                     3, multilabel=False, target_etype=("n0", "r1", "n1"),
+                                     10, multilabel=False, target_etype=("n0", "r1", "n1"),
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:782]: Why change this?

File: tests/unit-tests/test_gnn.py [hunk 3]
```diff
@@ -2471,6 +2512,448 @@ def check_call_gen_embs(skip_last_self_loop):
...
+                                      world_size=1)
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # get the test dummy distributed graph
+        _, part_config = generate_dummy_dist_graph(tmpdirname)
+        ep_data = GSgnnData(part_config=part_config,
+                            node_feat_field="feat",
+                            edge_feat_field="feat")
+
+    # Test case 0: normal case, model uses edge feature, and provides edge feature
+    #              for mini-batch inference
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:2534]: There can be another case: model uses edge feature feat0 but provides edge feature feat1 for mini-batch.
what will happen in this case?

File: tests/unit-tests/test_gnn.py [hunk 4]
```diff
@@ -2471,6 +2512,448 @@ def check_call_gen_embs(skip_last_self_loop):
...
+        ninput_feats = np_data.get_node_feats(input_nodes, nfeat_fields)
+        efeat_fields = dataloader2.edge_feat_fields
+        einput_feats_list = np_data.get_blocks_edge_feats(blocks, efeat_fields)
+
+        lbl = np_data.get_node_feats(seeds, 'label')
+        loss2 = model2(blocks, ninput_feats, einput_feats_list, lbl, input_nodes)
+        assert th.is_floating_point(loss2)
+
+    # Test case 2: abnormal case, edge model uses edge feature, and not provides edge feature
+    #              for mini-batch forward. Should Trigger an assertion error about the number
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:2636]:
```suggestion
    #             in forward. Should Trigger an assertion error about the number
```

File: tests/unit-tests/test_gnn.py [hunk 5]
```diff
@@ -2471,6 +2512,448 @@ def check_call_gen_embs(skip_last_self_loop):
...
+                                      rank=0,
+                                      world_size=1)
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # get the test dummy distributed graph
+        _, part_config = generate_dummy_dist_graph(tmpdirname)
+        np_data = GSgnnData(part_config=part_config,
+                            node_feat_field="feat",
+                            edge_feat_field="feat")
+
+    # Test case 0: normal case, node model uses edge feature, and provides edge feature
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:2590]: ditto

File: tests/unit-tests/test_gnn.py [hunk 6]
```diff
@@ -2471,6 +2512,448 @@ def check_call_gen_embs(skip_last_self_loop):
...
+        efeat_fields = dataloader1.edge_feat_fields
+        einput_feats_list = ep_data.get_blocks_edge_feats(blocks, efeat_fields)
+
+        label_field = dataloader1.label_field
+        lbl = ep_data.get_edge_feats({("n0", "r1", "n1"): [0, 1]}, label_field)
+        loss1 = model1(blocks, batch_graph, ninput_feats, einput_feats_list,
+                       None, lbl, input_nodes)
+        assert th.is_floating_point(loss1)
+
+    # Test case 1: normal case, edge model dose not use edge feature, and does not
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:2726]: ditto

File: tests/unit-tests/test_gnn.py [hunk 7]
```diff
@@ -2471,6 +2512,448 @@ def check_call_gen_embs(skip_last_self_loop):
...
+    #              of input features.
+    edge_feat_name = {('n0', 'r0', 'n1'): 'feat', ('n0', 'r1', 'n1'): 'feat'}
+    model3 = create_rgcn_edge_model(ep_data.g, num_ffn_layers=0, edge_feat_name=edge_feat_name)
+    target_idx = {("n0", "r1", "n1"): [0, 1]}
+    dataloader3 = GSgnnEdgeDataLoader(ep_data,
+                                      target_idx,
+                                      fanout=[-1, -1],
+                                      batch_size=10,
+                                      label_field='label',
+                                      node_feats='feat',
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:2761]:
```suggestion
                                      node_feats='feat',
                                      edge_feats=None,
```

Comment (classicsong) [reply tests/unit-tests/test_gnn.py:2761]: Let's explicitly set edge_feats to None.

File: tests/unit-tests/test_gnn.py [hunk 8]
```diff
@@ -2471,6 +2512,448 @@ def check_call_gen_embs(skip_last_self_loop):
...
+                                      rank=0,
+                                      world_size=1)
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # get the test dummy distributed graph
+        _, part_config = generate_dummy_dist_graph(tmpdirname)
+        ep_data = GSgnnData(part_config=part_config,
+                            node_feat_field="feat",
+                            edge_feat_field="feat")
+
+    # Test case 0: normal case, lp model uses edge feature, and provides edge feature
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:2821]: ditto

File: python/graphstorm/model/edge_gnn.py [hunk 9]
```diff
@@ -132,14 +132,20 @@ def forward(self, blocks, target_edges, node_feats, edge_feats, target_edge_feat
         labels, input_nodes=None):
         """ The forward function for edge prediction.
 
+        .. versionchanged:: 0.4.0
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/edge_gnn.py:135]: Yes, removed.

File: python/graphstorm/model/gnn.py [hunk 10]
```diff
@@ -864,37 +864,60 @@ def comput_input_embed(self, input_nodes, input_feats):
 
         return embs
 
-    def compute_embed_step(self, blocks, input_feats, input_nodes):
+    def compute_embed_step(self, blocks, input_nfeats, input_nodes, input_efeats_list=None):
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/gnn.py:867]: added.

File: python/graphstorm/model/gnn.py [hunk 11]
```diff
@@ -864,37 +864,60 @@ def comput_input_embed(self, input_nodes, input_feats):
...
+            node_embs = self.node_input_encoder(input_nfeats, input_nodes)
+            node_embs = {name: emb.to(device) for name, emb in node_embs.items()}
         else:
-            embs = input_feats
+            node_embs = input_nfeats
+        # For backward compatibility, set input edge feature list to be None
+        if input_efeats_list is None:
+            input_efeats_list = [{} for _ in range(len(blocks))]
+        if self.edge_input_encoder is not None:
+            edge_embs = self.edge_input_encoder(input_efeats_list)
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/gnn.py:905]: Changed both `node_embs` and `edge_embs`.

File: python/graphstorm/model/gnn.py [hunk 12]
```diff
@@ -1023,36 +1050,78 @@ def do_mini_batch_inference(model, data, batch_size=1024,
...
                 input_nodes = {data.g.ntypes[0]: input_nodes}
-            return {ntype: input_embeds[ntype][ids].to(device) \
-                    for ntype, ids in input_nodes.items()}
+            node_input_embs = {ntype: input_embeds[ntype][ids].to(device) \
+                               for ntype, ids in input_nodes.items()}
+            # get input edge features list
+            efeat_fields = data.edge_feat_field
+            if efeat_fields is not None:
+                if not isinstance(efeat_fields, dict):
+                    assert len(data.g.canonical_etypes) == 1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/gnn.py:1069]: added.

File: python/graphstorm/model/gnn.py [hunk 13]
```diff
@@ -1023,36 +1050,78 @@ def do_mini_batch_inference(model, data, batch_size=1024,
...
-            feats = prepare_batch_input(data.g, input_nodes, dev=device,
-                                        feat_field=data.node_feat_field)
-            return model.node_input_encoder(feats, input_nodes)
+            input_nfeats = prepare_batch_input(data.g, input_nodes, dev=device,
+                                               feat_field=data.node_feat_field)
+            # get input edge features list
+            efeat_fields = data.edge_feat_field
+            if efeat_fields is not None:
+                if not isinstance(efeat_fields, dict):
+                    assert len(data.g.canonical_etypes) == 1
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/gnn.py:1104]: added.

File: python/graphstorm/model/gnn_encoder_base.py [hunk 14]
```diff
@@ -101,6 +100,16 @@ def is_support_edge_feat(self):
...
 
+    def is_using_edge_feat(self):
+        """ Check if an instance of this class is using edge features.
+
+        If a GraphConvEncoder child class enables edge feature support, and is provided with
+        non-None ``edge_feat_name``, will return True. This method is for functions related
+        to trainers and inferrers, e.g., ``do_full_graph_inference()``, to determine if they
+        allow to use edge features.
+        """
+        return self.edge_feat_name is not None
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/gnn_encoder_base.py:111]: Because the `is_support_edge_feat()` directly do assert check, check it in this method will cause assertion error even when users do not want to use edge features.

File: tests/unit-tests/test_embed.py (hunk 1 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_embed.py:206]: No. It will be `('n0', 'r0', 'n1'): 2, ('n0', 'r1', 'n1'): 2`. This test setting is when `edge_input_layer` has weights, but the input is an empty dict list, the input layer still works.

File: tests/unit-tests/test_gnn.py (hunk 2 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_gnn.py:782]: Because the test data's label is 10 classes, not 3. The previous test code is wrong, but not be used in test cases.

Below is the test data code in `data_utils.py`.

```
    hetero_graph.nodes['n1'].data['label'] = th.randint(10, (hetero_graph.number_of_nodes('n1'), ))

    ......

    hetero_graph.edges['r1'].data['label'] = th.randint(10, (hetero_graph.number_of_edges('r1'), ))

```

File: tests/unit-tests/test_gnn.py (hunk 3 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_gnn.py:2534]: Add another test case, this will trigger an assert error. In addition, this case has been tested in `test_embed.py` line #281.

File: tests/unit-tests/test_gnn.py (hunk 5 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_gnn.py:2590]: Add another test case, this will trigger an assert error. In addition, this case has been tested in test_embed.py line https://github.com/awslabs/graphstorm/pull/281.

File: tests/unit-tests/test_gnn.py (hunk 6 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_gnn.py:2726]: Add another test case, this will trigger an assert error. In addition, this case has been tested in test_embed.py line https://github.com/awslabs/graphstorm/pull/281.

File: tests/unit-tests/test_gnn.py (hunk 8 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_gnn.py:2822]: Add another test case, this will trigger an assert error. In addition, this case has been tested in test_embed.py line https://github.com/awslabs/graphstorm/pull/281.

File: tests/unit-tests/test_gnn.py [hunk 15]
```diff
@@ -2546,18 +2546,18 @@ def test_edge_model_inference_with_edge_feats():
...
 
-    # Test case 1: normal case, model uses edge feature, and do full graph inference. 
+    # Test case 1: abnormal case, model uses edge feature, and do full graph inference. 
     #              This should trigger an assertion error, showing an edge feature supported
     #              GNN model cannot do full graph inference
     with assert_raises(AssertionError):
         do_full_graph_inference(model1, ep_data)
 
-    # Test case 2: normal case, model does not use edge feature, so can do both mini-batch and
+    # Test case 2: abnormal case, model does not use edge feature, so can do both mini-batch and
```
Comment (classicsong) [inline tests/unit-tests/test_gnn.py:2555]: Why this is abnormal?

**Review by classicsong — APPROVED** [review]

File: tests/unit-tests/test_gnn.py (hunk 15 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_gnn.py:2555]: revert back. It is a normal case.

File: python/graphstorm/model/gnn_encoder_base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/gnn_encoder_base.py:112]: updated the doc string to avoid confusion.
